### DISCUSSION TITLE
Is class_name on scripts a hard requirement?

### DISCUSSION THREAD

**Discussion by maximinus** [MikeSchulze/gdUnit3#361] [General]
Started testing GDUnit this week and it took me an age to get the first test running, because I could not find a way to reference the object I was testing. After really carefully looking at the code images in the docs, I finally discovered that I had to add class_name to the gdscript node I was testing, and then I could reference that object in the test.

I think this needs to be made a bit clearer in the tutorial because it nearly made me delete the addon.

Secondly, is this a hard requirement? I'm not necessarily against adding class_name, however it's something I've typically not done with my gdscript code before.

**Comment by MikeSchulze** [comment]
Hello and happy new year.
It sucks to hear that you are struggling to take your first test.

I don't now what you mean with `it is required to use class_name`
The test itself must be extend only from `GdUnitTestSuite` and the objects to tests are also not required to using class_name.

as simplest example for a test
```gdscript
extends GdUnitTestSuite

func test_foo():
	var obj = auto_free(load("res://ObjectToTest.gd").new())
	assert_that(obj.foo()).is_equal("foo")
```
ObjectToTest.gd
```gdscript
extends Node


func foo() -> String:
	return "foo"

```

The easiest way to use the Buildin function is to open the context menu above the function you want to test and select "Create Test".
Can you please give me more information/details about their problem.

**Reply by maximinus** [reply]
Thanks for the quick reply. Here's what I mean. I start a new scene and add a script. In the script, let's say I place this bit of code:

```
func i_am_false():
	return false
```

So i want to add a test for this, so I right-click the function and select "Create Test", which creates a new gdscript file for me, and a mock test.

From here, there is no way to actually reference the function i want to test. I can't call i_am_false directly, nor is any instanced object available to me.

The solution was to go to the original file and add this to the top:

```
class_name Player
```

And then change the test to

```
func test_i_am_false() -> void:
	var obj = Player.new()
	assert_bool(obj.i_am_false()).is_false()
```

Finally, as an extra, I still had issues because the objects `_ready()` is not called, so I actually ended up adding:

```
func before():
	player = auto_free(Player.new())
	player._ready()
```

Neither of these are bugs as such, but they made the tutorial very hard for me. The whole of your first tutorial will not work if the class_name is missing from the first file, and it took me too long to notice this. A simple extra line to say this would be great!

**Comment by MikeSchulze** [comment]
> Neither of these are bugs as such, but they made the tutorial very hard for me. The whole of your first tutorial will not work if the class_name is missing from the first file, and it took me too long to notice this. A simple extra line to say this would be great!

If I understand you correctly, you are a beginner in the Godot engine.
Using GdUnit assumes you understand how the Godot engine works, but yes, I can extend the startup tutorial to account for such pitfalls 👍 



In Godot there are two ways to instantiate an object.

You have to load a class via the classpath, or if you have defined a `class_name` you can instantiate it via the class name.
e.g.
`Player.new()`  needs a defined `class_name Player`
or
`load("res://Player.gd).new()`

Regarding your second problem, the `_ready()` function is only called when you add a node to the scene tree, otherwise you have to call it manually.
That is the way Godot works.

**Reply by maximinus** [reply]
> If I understand you correctly, you are a beginner in the Godot engine.

I've been using it since about 2.1

I think perhaps my assumption was that there would be some magic happening so that the function I want to test was actually ready to test. If you follow the tutorial literally, then, as I discovered, gdunit appears not to work and it's not immediately obvious what the issue is. Of course it DOES work, but I was very close to giving up at the first hurdle.

Perhaps you could add some extra code to the XXXTest.gd file created, i.e. if you were adding a test to a file `Foo.gd`:

```
var foo

func begin():
    foo = load('path/to/Foo.gd').new()
```

Although that's only a couple of minutes of thinking put into that.

> The _ready() function is only called when you add a node to the scene tree

Perhaps mention this in same tutorial? It's not certain what state the test code is run under, and a lot of people are likely to use `_ready()` as a form of constructor - even the official docs state "Usually used for initialization"

**Reply by MikeSchulze** [reply]
Hi @maximinus good points, thanks for your feedback.
I will add this missing information to the documentation.

I am always open to improvements and suggestions.
Thank you very much!
